pkg/kv/kvserver/kvstorage/init.go

```diff
@@ -356,37 +398,103 @@ func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineR
 		return nil, err
 	}
 
+	// INVARIANT: all replicas have a persisted full replicaID (i.e. a "replicaID from disk").
+	//
+	// This invariant is true for replicas created in 22.2. Without further action, it
+	// maybe be violated for clusters that originated before 22.2.
+	//
+	// loadFullReplicaIDsFromDisk returns a zero ReplicaID when none is found.
+	//
+	// In this method, we migrate to always satisfy the invariant:
+	//
+	// 1. for initialized replicas, use ReplicaID from RangeDescriptor as RaftReplicaID
+	// 2. remove uninitialized replicas lacking a RaftReplicaID.
+	//
+	// The migration can be removed when the KV host cluster MinSupportedVersion
+	// matches or exceeds 23.1 (i.e. once we know that a store has definitely
+	// started up on 23.1 at least once).
+
 	allM, err := loadFullReplicaIDsFromDisk(ctx, eng)
 	if err != nil {
 		return nil, err
 	}
 
-	for id := range initM {
-		if _, ok := allM[id]; !ok {
-			// INVARIANT: all replicas have a persisted full replicaID (i.e. a "replicaID from disk").
-			//
-			// This invariant is true for replicas created in 22.2, but no migration
-			// was ever written. So we backfill the replicaID here (as of 23.1) and
-			// remove this code in the future (the follow-up release, assuming it is
-			// forced to migrate through 23.1, otherwise later).
-			if buildutil.CrdbTestBuild {
-				return nil, errors.AssertionFailedf("%s has no persisted replicaID", initM[id])
-			}
-			if err := logstore.NewStateLoader(id.RangeID).SetRaftReplicaID(ctx, eng, id.ReplicaID); err != nil {
-				return nil, errors.Wrapf(err, "backfilling replicaID for r%d", id.RangeID)
-			}
-			log.Eventf(ctx, "backfilled replicaID for %s", id)
+	for id := range allM {
+		if id.ReplicaID != 0 {
+			// RaftReplicaID was present, no need to backfill.
+			continue
 		}
-		// `allM` will be our map of uninitialized replicas.
+
+	}
+
+	for id := range initM {
```

Comment:
Not reviewed today below this line. Will continue tomorrow, but also would like to see if this PR can be split: (1) complete backfilling ReplicaIDs, (2) some time later also look at removing ReplicaIDs that we don't need (rationale in the prev. comment).

Reply:
I rewrote according to your suggestion and I think the result is orders of magnitude clearer. I tend to think splitting it at this point is more busy work for me than benefit for you but let me know (once I've pinged) and I can still tease it apart after.